### DISCUSSION TITLE
BDEW profile for AS4

### DISCUSSION THREAD

**Discussion by deupalmah** [phax/phase4#105] [Q&A]
Hi there, 
does anyone know if there is a ready to use BDEW profile for phase 4 ?
Maybe in development and to be shared soon?

**Comment by phax** [comment]
Can you please elaborate what BDEW is and eventually provide some links here?

**Comment by deupalmah** [comment]
The BDEW is the largest energy industry association in Germany. They specified an AS4 profile for exchanging energy related information.
https://www.bundesnetzagentur.de/DE/Beschlusskammern/1_GZ/BK6-GZ/2021/BK6-21-282/Mitteilung01/Regelung_Uebertragungsweg.pdf?__blob=publicationFile&v=1

**Reply by phax** [reply]
Thanks for the clarification. I just knew it as "edi@energy". Currently this profile is not usable out of the box with phase4, especially because a certain aspect of the security requirements is currently not supported by the underlying WSS4J library. But I know some people are working on trying to use phase4 for BDEW. hth

**Reply by deupalmah** [reply]
What are the security requirements which are not supported? Can you give me a more specific answer, please?
WSSecurity is needed and offered by wss4j, what is the difference between the provided profiles?

here i have some further questions:
Do you think it is difficult to replace wss4j here in the project? Is it there too much to do?
Are the projects of the people you know  a free fork of phase4?

Thank you in advance

**Reply by phax** [reply]
Well, one the hand side they require the usage of the "BrainpoolP256r1" EC curve - this can be handled with BouncyCastle.
But there was some requirement on the fields of XML security that was currently not suported by WSS4J - I need to check my mail archive (will do later).

I don't know what the status of the guys is, that work on BDEW support. But I brought this thread to their attention. So its up to them to decide if they want to communicate or not.

Replacing wss4j might be an option, but I have never thought about this. Are you aware of an alternative OpenSource implementation?

**Reply by phax** [reply]
I think the problem is, that KeyDerivationMethod is currently not supported by WSS4J (still need to look it up).
To my understanding Holodeck, Domibus, Oxalis and phase4 all rely on wss4j.

**Reply by phax** [reply]
So I checked, and the probem is, that WSS4J does not support Key Agreement - hth

**Reply by deupalmah** [reply]
thank you very much for your effort.

**Reply by Stefan4112** [reply]
Hi, we are currently evaluating and testing phase4. The Key Agreement is a real problem, because it is difficult to implement. Here is a link to the feature request in santuario which is used by WSS4J:
https://issues.apache.org/jira/browse/SANTUARIO-511

Also we do not understand why it is used for File Encryption... very unusual

We thought about replacing WSS4J, too. But everyone seems to use it.

**Reply by deupalmah** [reply]
Stefan4112 you are right, nearly every SOAP Framework relies on WSS4J for encryption.
The feature request is from 2019. As i can see you asked for support. Did you get any reply or help there?

**Reply by Stefan4112** [reply]
We contacted the developers (in case we found an email), but they could not help us. They have no time for this feature.

**Reply by Stefan4112** [reply]
I posted an update at the Santuario issue: Seems like Key Agreement is not a problem. BouncyCastle can calculate it and the result can be used as a parameter in WSS4J.

**Reply by phax** [reply]
@Stefan4112  thank you for the update - that really sounds promising!!!

**Reply by phax** [reply]
The https://issues.apache.org/jira/browse/SANTUARIO-511 issue was resolved in release 3.0.4

**Comment by deupalmah** [comment]
so far I'm watching and evaluating. There are of course other open source AS4 applications. But my focus is currently on phase4 as it is well maintained and you are replying quickly. 
i am wainting for your futher response out of your mail archive :)

**Comment by francbohuslav** [comment]
I was also asked to check support for BDEW in Phase4. Please give us any solution even if it is not solved by Phase4. Our team is also trying to find any suitable solution and I will give you feedback if we find anything helpful.

**Reply by phax** [reply]
From @Stefan4112

>  The Key Agreement is a real problem, because it is difficult to implement. Here is a link to the feature request in santuario which is used by WSS4J:
> https://issues.apache.org/jira/browse/SANTUARIO-511
> Also we do not understand why it is used for File Encryption... very unusual
> We thought about replacing WSS4J, too. But everyone seems to use it.

**Reply by Stefan4112** [reply]
I posted an update at the Santuario issue: Seems like Key Agreement is not a problem. BouncyCastle can calculate it and the result can be used as a parameter in WSS4J.

**Comment by francbohuslav** [comment]
Information from Santuario JIRA leads to Slack https://phoss.slack.com/ and seems that @phax participates somehow on that. Can I humbly ask for adding **unicornsystems.eu** to allowed domains of that slack server to see channel **#as4-de-bdew**?

**Reply by phax** [reply]
@francbohuslav yes, I will send you an invite.
Edit: the domain was added

**Reply by deupalmah** [reply]
Hi @phax 
please add me too (with gmail account when possible)

**Reply by phax** [reply]
@deupalmah I can't find your email address (or I can't link it). Please send it to me - you can find my email address in the pom.xml at https://github.com/phax/phase4/blob/master/pom.xml#L60

**Reply by francbohuslav** [reply]
@phax Thanks, I am in

**Reply by phax** [reply]
@francbohuslav can't see you yet

**Reply by CJSmith-0141** [reply]
@phax Could I get added with my work email `connor.smith1@octoenergy.com`? I am working on integrating phase4 and would appreciate the camaraderie at the least.

**Comment by sopgreg** [comment]
@phax can you please also add **@soptim.de** to the list of allowed domains for the channel **#as4-de-bdew**? As a software vendor for the DACH energy industry, we would also like to participate and contribute to the project. Thanks!

**Reply by phax** [reply]
Done. Enjoy :)

**Reply by sopgreg** [reply]
Thanks a lot!

**Comment by CJSmith-0141** [comment]
I'm in the process of trying to set up phase4 to work with my application written in Scala. I've already found that I have needed to downgrade to a version 11 JVM from a version 17, which is fine.

Where I'm getting stuck now is I don't know how to get my outgoing message to be encrypted. Following is a snippet of the logs I'm seeing . . .

```
[info] 2023:271:15:32:41.923 [io-compute-blocker-0] INFO  c.h.p.m.MetaAS4Manager MetaAS4Manager is initialized with in-memory data structures
[info] 2023:271:15:32:41.926 [io-compute-blocker-0] INFO  c.h.p.m.MetaAS4Manager Creating AS4 managers using factory class com.helger.phase4.mgr.AS4ManagerFactoryInMemory
[info] 2023:271:15:32:41.934 [io-compute-blocker-0] INFO  c.h.p.p.AS4ProfileManager 1 AS4 profile is registered
[info] 2023:271:15:32:41.935 [io-compute-blocker-0] INFO  c.h.p.m.MetaAS4Manager MetaAS4Manager was initialized
[info] 2023:271:15:32:41.967 [io-compute-blocker-0] INFO  c.h.p.s.AS4BidirectionalClientHelper Sending AS4 UserMessage to 'http://localhost:18141/as4' with max. 1 retries
[info] 2023:271:15:32:41.968 [io-compute-blocker-0] WARN  c.h.p.c.AS4ClientUserMessage Crypt parameters have an algorithm defined but neither an alias nor a certificate was provided. Therefore encryption is not enabled.
[info] 2023:271:15:32:42.231 [io-compute-blocker-0] INFO  c.h.p.w.WSSConfigManager None of the WSSConfig Security Providers is already installed - doing it now
[info] 2023:271:15:32:42.256 [io-compute-blocker-0] INFO  c.h.p.m.c.AS4Signer Now signing AS4 message [SOAP_12]. KeyIdentifierType=BST_DIRECT_REFERENCE; KeyAlias=1; SignAlgo=http://www.w3.org/2001/04/xmldsig-more#ecdsa-sha256; DigestAlgo=http://www.w3.org/2001/04/xmlenc#sha256; C14NAlgo=http://www.w3.org/2001/10/xml-exc-c14n#
[info] 2023:271:15:32:42.297 [io-compute-blocker-0] WARN  o.a.j.x.d.i.d.DOMReference The input bytes to the digest operation are null. This may be due to a problem with the Reference URI or its Transforms.
[info] 2023:271:15:32:42.349 [io-compute-blocker-0] INFO  c.h.p.h.BasicHttpPoster Starting to transmit AS4 Message to 'http://localhost:18141/as4'
[info] 2023:271:15:32:44.429 [io-compute-blocker-0] INFO  c.h.p.h.BasicHttpPoster Finished transmitting AS4 Message to 'http://localhost:18141/as4' after 2074 ms
[info] 2023:271:15:32:44.430 [io-compute-blocker-0] INFO  c.h.p.s.AS4BidirectionalClientHelper Successfully transmitted AS4 UserMessage with message ID '7be5743b-78bd-4c5d-a927-04628c34cf6d@phase4' to 'http://localhost:18141/as4'
```
specifically this line is of interest to me 
```
[info] 2023:271:15:32:41.968 [io-compute-blocker-0] WARN  c.h.p.c.AS4ClientUserMessage Crypt parameters have an algorithm defined but neither an alias nor a certificate was provided. Therefore encryption is not enabled.
```
Here is a minimized version of the code I'm using to initialize the send (apologies for the Scala-ness, hopefully it's legible enough)
```scala
for {
  payloadParams: BDEWPayloadParams <- IO.pure {
    val params = new BDEWPayloadParams()
    params.setDocumentType("DT1")
    params.setDocumentDate(PDTFactory.getCurrentLocalDate)
    params.setDocumentNumber(456)
    params.setFulfillmentDate(
      PDTFactory.getCurrentLocalDate.minusMonths(2)
    )
    params.setSubjectPartyId("AS4-Receiver")
    params.setSubjectPartyRole("Role1")
    params
  }
  httpClientSettings: Phase4BDEWHttpClientSettings <- IO.pure {
    val settings = new Phase4BDEWHttpClientSettings()
    settings.setUserAgent("kraken-siphon")
    settings.setSSLContextTrustAll()
    settings
  }
  encryptionProps <- IO.pure {
    val factory    = AS4CryptoFactoryProperties.getDefaultInstance
    val properties = factory.cryptoProperties()
    properties.setKeyAlias("1")
    properties
  }
  send <- IO.blocking {
    Phase4BDEWSender
      .builder()
      .encryptionKeyIdentifierType(
        ECryptoKeyIdentifierType.X509_KEY_IDENTIFIER
      )
      .signingKeyIdentifierType(
        ECryptoKeyIdentifierType.BST_DIRECT_REFERENCE
      )
      .fromPartyID("AS4-Sender")
      .fromRole(
        CAS4.DEFAULT_INITIATOR_URL
      )
      .toPartyID("AS4-Receiver")
      .toRole(
        CAS4.DEFAULT_RESPONDER_URL
      )
      .endpointURL("http://localhost:18141/as4")
      .service(BDEWPMode.SERVICE_MARKTPROZESSE)
      .action(BDEWPMode.ACTION_DEFAULT)
      .httpClientFactory(httpClientSettings)
      .cryptoFactoryCrypt(
        new AS4CryptoFactoryProperties(encryptionProps)
      )
      .payload(
        AS4OutgoingAttachment
          .builder()
          .data("Hello World!".getBytes())
          .compressionGZIP()
          .mimeTypeXML()
          .charset(StandardCharsets.UTF_8),
        payloadParams
      )
      .sendMessageAndCheckForReceipt()
  }
  _ <- IO.delay(
    logger.warn(send.toString)
  )
} yield ()
```
Here is a redacted version of my `phase4.properties`
```
org.apache.wss4j.crypto.merlin.keystore.type=PKCS12
org.apache.wss4j.crypto.merlin.keystore.file=/Users/csmith/siphon/siphon.pfx
org.apache.wss4j.crypto.merlin.keystore.password=<redacted>
org.apache.wss4j.crypto.merlin.keystore.private.password=<redacted>
org.apache.wss4j.crypto.merlin.keystore.alias=1
org.apache.wss4j.crypto.merlin.keystore.provider=BC
org.apache.wss4j.crypto.merlin.cert.provider=BC
org.apache.wss4j.crypto.merlin.truststore.provider=BC
global.debug=true
global.production=false
global.datapath=/Users/csmith/siphon/datapath
phase4.profile=BDEW
```

**Reply by phax** [reply]
Hi, in general BDEW is NOT fully supported by phase4, because the key exchange is not fully implemented in the underlying WSS4J library. So there is a good chunk of encryption code that you need to add manually - sorry. I will also highlight that in the docs.

Java 17 should work. Java 11 is just the minimum Java version.

Your problem is denoted in the log messages as
> Crypt parameters have an algorithm defined but neither an alias nor a certificate was provided. Therefore encryption is not enabled.

You need to add the receiverCertificate to the `cryptParams` e.g. via this Java snippet on the builder:
```java
.withCryptParams (cp -> cp.setCertificate (aReceivertCert))
```

**Reply by beth-soptim** [reply]
Hi,

it seems that you want to use Phase4 to participate in AS4 for the German electricity market. You will have to extend Phase4/WSS4J (see PHs comment above).

Relevant documents (by BDEW/BSI/Root-CA/BNetzA) are (at least):

* TR-03109-1, Anforderungen an die Interoperabilität der Kommunikationseinheit eines intelligenten Messsystems
* TR-03109-4, Smart Metering PKI – Public Key Infrastructure für Smart Meter Gateways.
* TR-03116-3, Kryptographische Vorgaben für Projekte der Bundesregierung, Teil 3: Intelligente Messsysteme.
* SM-PKI-CP: Certificate Policy der Smart Metering PKI
* KLC-SR: Key Lifecycle Security Requirements
* TR-02102-1, Kryptographische Verfahren: Empfehlungen und Schlüssellängen.
* TR-03116-4, Kryptographische Vorgaben für Projekte der Bundesregierung, Teil 4: Kommunikationsverfahren in Anwendungen.
* Regelungen zum Übertragungsweg für AS4 2.0
* BDEW AS4-Profil: AS4-Nutzungsprofil zum Datenaustausch für regulierte Prozesse in der Energiewirtschaft
* BNetzA BK6-21-282
* BNetzA BK6-18-032
* BNetzA BK7-16-142

Most of these documents are only available in German.

There are very special requirements concerning TLS (Brainpool curves, session resumption, HMAC shortening, ...) and XML signing/encryption.

Also, you will not be able to use standard tools like OpenSSL for creating correct Smart Metering PKI certificate requests (see documents above). You will have to write or buy specialized tools.

Are you a member of https://phoss.slack.com/?

**Reply by CJSmith-0141** [reply]
No, I am not in that slack. Can I be added? `connor.smith1@octoenergy.com`

**Reply by beth-soptim** [reply]
I think @phax has to do this.

**Reply by phax** [reply]
@CJSmith-0141 done. When joined, please use #as4-de-bdew channel. CU

**Reply by marekgregor** [reply]
@phax could you please also add me (`marek.gregor@gmail.com`) as slack member of https://phoss.slack.com/. Thanks a lot.

**Comment by vbrandl** [comment]
https://issues.apache.org/jira/browse/SANTUARIO-511 has been implemented in xmlsec 3.0.4. Would it help for the BDEW profile, if phase4 updates this dependency?

**Reply by phax** [reply]
Sure - will do it anyway for 2.7.5 - thanks for the headsup though

**Reply by sbarlabanov** [reply]
There is still a problem with xmlsec (at least 3.0.4 - did not check 4.x yet): they added support for EDHS-ES key agreement but they do not support brainpool curves required by BDEW profile. The curves themself are supported by Bouncy Castle but xmlsec fails to work with keys based on brainpool curve. xmlsec 3.0.4 supports only NIST curves. Unfortenutelly BrainpoolP256r1 is the only allowed curve in BDEW profile. We are currently looking for ways to add the curves support to xmlsec.

**Reply by sbarlabanov** [reply]
There is already a PR to add brainpool curves support to xmlsec: https://github.com/apache/santuario-xml-security-java/pull/292

**Reply by phax** [reply]
That's great news. And https://github.com/apache/santuario-xml-security-java/pull/293 is the same PR on the correct branch

**Reply by Mumitrol** [reply]
Is the release of version 2.7.5 already scheduled?

**Reply by phax** [reply]
It is planned for today or tomorrow

**Reply by phax** [reply]
version 2.7.5 was released - propagation to Maven Central may take a few minutes

**Comment by leepriest123** [comment]
@phax would it be possible to add my address lee.priest@bradyplc.com to the slack channel #as4-de-bdew? We are working on this AS4 with the bdew profile and would appreciate the opportunity to share any wisdom from others in the channel.

**Comment by phax** [comment]
There still seems to be an issue with using wss4j 3.0.5 for BDEW.
See https://issues.apache.org/jira/browse/WSS-714 and https://github.com/apache/ws-wss4j/pull/418 for a potential remediation

**Reply by philippnagel** [reply]
@phax please add `mako365.com` as an allowed domain to the slack. 

We are an energy data and process platform and as part of this use AS4 in several ways.

**Reply by phax** [reply]
Sorry, I don't know how to only add the domain. Send me an email (see pom.xml) with your email address and than I can also whitelist the domain
